### ISSUE TITLE
XElement Namespace Handling

### ISSUE THREAD

**Issue by Inspyro** [dotnet/corefx#7172]
I found a problem with the way XElement handles namespaces. If You look at the following code (executed with .Net Core 1.0.0-rc2-23811):

```C#
var x1 = new XElement("{http://example.com}root");
Console.WriteLine(x1);     
            
var x2 = XElement.Parse(x1.ToString());
Console.WriteLine(x2);
            
Console.WriteLine($"Attributes of x1: '{string.Join(", ", x1.Attributes())}'");
Console.WriteLine($"Attributes of x2: '{string.Join(", ", x2.Attributes())}'");
```

The output is as follows:
```xml
<root xmlns="http://example.com" />
<root xmlns="http://example.com" />
Attributes of x1: ''
Attributes of x2: 'xmlns="http://example.com"'
``` 

To me this behaviour seems strange / **inconsistent**, as one **cannot rely** that the xmnls attribute is represented as **xmlns attribute in the memory model**. Some Frameworks (e.g. Json.NET) however require the xmlns attribute to exist in the memory model, which leads to strange results. 

My suggestion is: 
- Either always add the xmlns attribute when using the syntax I wrote above automatically.
- Or never add the xmnls attribute (disallow adding it in the api) - also not during parsing xml, but make it implicit (accessible via Name.Namespace)

So basically keep the Name.NamespaceName and the xmnls attribute in sync.

**Comment by Inspyro** [comment]
Have a look at this Json.Net issue:
https://github.com/JamesNK/Newtonsoft.Json/issues/853

**Comment by fschmied** [comment]
I think there is a good reason for not just always adding an `xmlns` `XAttribute` whenever a `Name.NamespaceName` is present: when a child `XElement` with a given default namespace has a parent with a corresponding `xmlns` attribute, it's "correct" that the child does _not_ have the `xmlns` atrribute, while it still has a `Name.NamespaceName`. So the object model really needs to support this in-between situation, where neither an inherited (from the parent element), nor an explicitly declared attribute are present (yet).

However, the situation really gets confusing when one calls `ToString`, where the `xmlns` attribute is suddenly inserted, even though it isn't there in the object model. I guess, the `ToString` method is between a rock and a hard place because that `XElement` state (no inherited and no explictly declared `xmlns` attribute, but still a `Name.NamespaceName`) cannot be expressed as an XML string.

In my opinion, `XElement.ToString` should really throw a `NotSupportedException` in that case rather than just adding the attribute. However, @Inspyro's second option (not expressing the `xmlns` as an `XAttribute` at all, just make it a "projection" of the `Name.Namespace` and the parent element) might be the even better option.

**Comment by JonHanna** [comment]
> cannot rely that the xmnls attribute is represented as xmlns attribute in the memory model

In which situations would one want to look to the attribute itself, rather than the namespace for the relevant prefix (or default namespace for no prefix)?

**Comment by Inspyro** [comment]
@JonHanna 
For me the api seems to suggest that you can also check the namespace via xmnls attribute from the attributes list, which works perfectly fine, as long as the xml was parsed - e.g. from a file or from a string.
But as soon as I use the fancy api for defining the xmnls namespace in memory, this does not work anymore.

This probably also allured the Json.Net devs into doing this check for XML to JSON serialization. (see: https://github.com/JamesNK/Newtonsoft.Json/issues/853)  

And I assume there could be more such frameworks out there that do this check in an incomplete way. So adapting the dotnet api could improve future frameworks that build upon it.

**Comment by JonHanna** [comment]
Consider:

```XML
<a xmlns="http://example.net/">
  <b />
</a>
```

The `<b />` element is clearly in the `http://example.net/` namespace, but the tag doesn't have that attribute.

Now, let's forget about `XElement` specifically, and just think generally about something that serialises XML elements into XML documents.

If we were to serialise the `<a />` element, including its children, then we'd expect the output to be much as the example above.

If we were to serialise the `<b/>` element alone, then we'd expect it to be output as:

```XML
<b xmlns="http://example.net/"/>
```

And then it would have that attribute.

All of which seems consistent with the behaviour you've described.

(I can see an argument for never reflecting XMLNS attributes as attribute items in an object model at all, but that would break things).

**Comment by fschmied** [comment]
@JonHanna In my opinion, the state of an `XElement` that has a `Name.NamespaceName`, but neither a parent element nor an `xmlns` attribute, is inconsistent (or unfinished) with regards to XML. In other words, there is no 1:1 representation for that `XElement` state in XML. Do you concur?

If so, there are two questions:

- Does `XElement` need to represent this inconsistent or unfinished state? There are alternatives, such as @Inspyro's second suggestion.
- If one accepts that `XElement` may represent unfinished states, does something that serializes XML elements into XML strings (I paraphrased your expression a little) need to support these unfinished states? I think that no, it could, for example, throw an exception stating that the `XElement` isn't ready for translating into a string yet..

Due to backwards compatibility, it's of course extremely unlikely that this behavior of `XElement` be changed. However, I think a warning in the documentation of `ToString` (and the explicit conversion `string`) stating that the operation may add information not present in the `XElement`, such as `xmlns` attributes, would be sensible. To me, at least, it was surprising that `XElement.Parse(xelement.ToString())` would yield a structurally different `XElement` than the original one.

**Comment by JonHanna** [comment]
>  In my opinion, the state of an XElement that has a Name.NamespaceName, but neither a parent element nor an xmlns attribute, is inconsistent (or unfinished) with regards to XML. In other words, there is no 1:1 representation for that XElement state in XML. Do you concur?

No. It has a local-name a (possibly default) prefix, and a namespace. It's consistent with the XML data model.

**Comment by karelz** [comment]
As discussed above, we won't change it due to compat. Please file a new doc bug if you think it will help.
